Fix synced pattern editing in write mode and refactor block editing mode to reducer

talldan opened the pull request:
## What?
Supercedes some previous attempts #66919, #66949, and #65408!
Fixes #66424

In Write Mode, parts of synced patterns that are supposed to be non-editable are editable. This PR fixes the problem, plus undertakes some refactoring of how block editing mode works due to some performance caused by so much logic being in the `getBlockEditingMode` selector.

## Why?
This bug happens because Write Mode overrides the blockEditingMode of a synced pattern's inner blocks.

Pattern blocks try calling `setBlockEditingMode` for child blocks imperatively (in the pattern block's edit function). This does set the correct block editing modes in the block editor store, but when write mode is active the `getBlockEditingMode` selector returns early with a different value.

#65408 tried to fix this by updating the `getBlockEditingMode` selector, but it cause a performance hit. 

## How?
This PR fixes the issue by moving the pattern block editing modes to the block editor store, so that it can work harmoniously with the way Write Mode and Zoom Out mode determine the block editing mode.

The PR also undertakes a refactor of how block editing mode works for zoomed out and write mode to improve performance. The `getBlockEditingMode` select is called a lot (including recursively), so any code added there can cause performance regressions.

The summary of the changes is:
- Remove the zoom out and write mode handling from the `getBlockEditingMode` selector
- Introduce a new `withDerivedBlockEditingModes` higher order reducer, it calculates a derivedBlockEditingMode state for any actions that might require this is recomputed using a few other functions:
    - `getDerivedBlockEditingModesForTree` traverses a block tree calculating the block editing modes for all the blocks in that tree, returning them in a `Map`. This can be used for the entire block tree or just a part of the block tree.
    - `getDerivedBlockEditingModeForBlock` calculates the block editing mode for a single client id. Called by `getDerivedBlockEditingModesForTree` on each block in the tree.
    - `getDerivedBlockEditingModesUpdates` used for actions that add/remove a partial subset of blocks. It should only return a value when there are block editing mode changes, meaning the store state doesn't unnecessarily change.
    - Plus a few utility functions - `traverseBlockTree`, `getBlockTreeBlock`, `findParentInClientIdsList`, `hasBindings`.
    - Also adds a new `isContentBlock` private function to the blocks package.
    
It's quite a lot of code, so I'm open to suggestions on how to structure it.
    
## Testing Instructions
1. Create a synced pattern (Pattern A) with two paragraphs
2. Edit the pattern and enable overrides for one of the patterns
3. Insert the pattern into a template
4. Check that you can only edit the overridden paragraph, the other paragraph should be disabled
5. Enable Zoomed Out mode, you shouldn't be able to edit any of the pattern content, since zoomed out doesn't allow content editing
6. Leave Zoomed Out mode and enable Write mode, check that you can only edit the overridden paragraph in the pattern
7. Create a second pattern (Pattern B) similar to the first one, it should have a block that's overridden and a regular block
8. Edit Pattern A and insert Pattern B into it.
9. Insert Pattern A into a template again, or go back to where you inserted it before
10. Check that you can't edit any of Pattern B from the instance of Pattern A, it should all be disabled

In addition to this, give a good test to both zoomed out and write mode, move blocks around, insert and remove stuff and make sure there are no issues. 😄 

## Screenshots or screencast <!-- if applicable -->
#### Before

https://github.com/user-attachments/assets/49cd5420-aa00-4437-be43-1d1f16002514

#### After

https://github.com/user-attachments/assets/feca05d3-9fb7-404a-8e5a-406e27b4f023

talldan committed:
Handle RESET_ZOOM_LEVEL action

talldan committed:
Add higher order reducer for pattern block editing modes.

- Remove the prev. React effect for managing pattern block editing modes
- Implement higher order reducer in the block editor store ... it:
    - Tracks the clientIds of pattern blocks.
    - Uses the pattern block clientIds to manage block editing modes
      for pattern blocks and their inner blocks.
    - Updates both on any actions that change block lists.

talldan committed:
Add higher order reducer for block editing modes while section editing

talldan committed:
Bug fixes

talldan committed:
Avoid mutating `state` in new higher order reducers

talldan committed:
Try moving synced pattern client ids into a separate reducer

talldan committed:
Revert "Try moving synced pattern client ids into a separate reducer"

This reverts commit e1d6ca494e6f830c429aa5649327b4c26d77a974.

It doesn't really work, since reducer won't have access to `state.blocks.tree`.

talldan committed:
Try amalgamating the different derived block editing modes

talldan committed:
Fixes

talldan committed:
Fix synced patterns in write mode, unbound content blocks being editable

talldan committed:
Also update derived block editing mode on `REPLACE_INNER_BLOCKS

talldan committed:
Fix descending through controlled inner blocks

talldan committed:
Fix nested pattern handling - always process synced patterns in the reducer. Add special handling for only adding pattern block itself as content only when not in zoomed out or nav mode

talldan committed:
Zoomed out fixes - content should never be editable in zoomed out, even synced pattern overrides or when write mode is active

talldan committed:
Docs

talldan committed:
Add end to end test

talldan committed:
Remove navigation mode selector tests

talldan committed:
Fix partial mocking of blocks package

talldan committed:
Add test for isContentBlock

talldan committed:
Add unit tests

talldan committed:
Remove defaultBlockEditingMode concept

talldan committed:
Handle patterns that are outside sections in nav mode

talldan committed:
Remove comment

talldan committed:
Refactor to handle tree subsections

talldan committed:
Optimize each individual action

talldan committed:
Fixes and refinements

talldan committed:
Comments and renamings

talldan committed:
Remove test mocking - else test chokes trying to unlock privateApis

youknowriad commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1669,14 +1669,16 @@ export const setNavigationMode =
 			speak( __( 'You are currently in Design mode.' ) );
 		}
+
+		dispatch( { type: 'SET_EDITOR_MODE', mode } );
```
This is the one that I'm not sure about because some people could call `registry.dispatch( preferencesStore ).set( 'core', 'editorTool', mode );` directly. Maybe it's fine I don't know

talldan replied:
Yeah, that's a good point, though I guess the spoken messages also won't work when changing the mode that way. Some options:
- (I think you mentioned this one) Always calculate the block editing modes for write mode even when in design mode, then the selector can choose the right one. This'd be a shame as I expect it'd cause a performance hit.
- Add a listener (probably an effect) that reacts to the preference change and sets the mode in the block editor store.
- Make the preference key private so that it can only be called internally. I think this preference has already shipped in the plugin, so it would need some back compat wrangling. Possibly my preferred solution though.

youknowriad replied:
Also, what happens on "init"? which mode is used?

I may be leaning towards the first option personally to decouple the preference setting entirely from the computation of the mode.

talldan replied:
> Also, what happens on "init"? which mode is used?

It uses whatever value the preferences store has at the time of computation. I guess there could be a concern if the preferences store isn't ready yet for some reason.

talldan replied:
Apart from resolving this question, the rest of the PR is pretty finalized (I've added lots of tests).

I'll have a go at your suggestion, I think it should be pretty easy to implement, we'll probably just need an extra state property for `navModeBlockEditingModes` or something. One thing on my mind is preventing this from being calculated when write mode isn't an option (I think in the post editor it's not an option), as that would be unnecessary.

youknowriad commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -3046,68 +3038,8 @@ export const getBlockEditingMode = createRegistrySelector(
-				const isContent = hasContentRoleAttribute( name );
-
-				return isContent ? 'contentOnly' : 'disabled';
+			if ( state.derivedBlockEditingModes?.has( clientId ) ) {
```
In which case, there's no "derived block editing mode"?

talldan replied:
`derivedBlockEditingModes` is only storing `contentOnly` or `disabled` values at the moment that are related to zoomed out, write mode or synced patterns.

I think it'd be possible to move the template locking code from the `getBlockEditingModes` selector to the reducer as well in the future, then `getBlockEditingModes` would be doing very little.

youknowriad commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -112,6 +112,7 @@ function getEnabledClientIdsTreeUnmemoized( state, rootClientId ) {
 export const getEnabledClientIdsTree = createRegistrySelector( ( select ) =>
 	createSelector( getEnabledClientIdsTreeUnmemoized, ( state ) => [
```
I think some of the dependencies here can probably be removed because they're not used in the selector anymore. But I don't expect that it impacts anything.

talldan replied:
Good call, I'll take a look!

youknowriad commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -2131,6 +2140,524 @@ const combinedReducers = combineReducers( {
+function getDerivedBlockEditingModeForBlock(
+	state,
+	clientId,
+	{
```
What's the difference between this argument and all the content that is already in `state`?

talldan replied:
It's mostly stuff that's already been retrieved or calculated by the calling function, so it's an optimization to pass it in rather than determine it again for every block.

In an earlier iteration this function was called from a couple of different places, but now it's only called from `getDerivedBlockEditingModesForTree` so I could inline it there.

It'd make a small dent in the line of code changed in this PR 😆

draganescu replied:
They seem to need to be derrived in various ways since they're not preset in this state?

youknowriad replied:
If the optimization is helpful, I don't mind it, but maybe at least at a comment.

talldan replied:
I've inlined this code now so no comment needed.

youknowriad reviewed (commented):
I didn't go deep in the algorithm review but the change itself feels right to me. I think it's we need some testing around it (maybe we already have)

draganescu commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -2131,6 +2140,524 @@ const combinedReducers = combineReducers( {
+		if (
+			isZoomedOut ||
+			! findParentInClientIdsList( state, clientId, sectionClientIds )
+		) {
```
This means if zoom out always set all blocks which are not sections diabled or if nav mode, which unintuitively now is actually write mode, then set to disabled all blocks that don't have section parents. It's pretty intricate.

draganescu commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -2131,6 +2140,524 @@ const combinedReducers = combineReducers( {
+			return 'disabled';
+		}
+	}
+}
```
Am I reading this wrong or we don't treat content locking parents? Seems we have when zoom out, when write mode, and when design mode and synced pattern.

draganescu replied:
Oh answered in https://github.com/WordPress/gutenberg/pull/67026/files?diff=unified&w=1#r1849986579 - we're specifically NOT dealing with content locking. At least for now. I read this as a sweeping change. It's just one step.

draganescu reviewed (commented):
This works pretty great already and the removals of effects plus centralising the decision makes everything easier to grasp. I wish we'd also handle the content locking in the same place.

talldan committed:
Rework reducer unit tests and add more cases

talldan committed:
Add more tests

ramonjd commented:
I'm curious to know if I'm testing this correctly or if what I'm seeing is expected.

When editing a post with the template preview off, the write/design mode drop down is not visible.

There's functionality in the post editor, for example, [in the block settings dropdown](https://github.com/WordPress/gutenberg/blob/6f69a3254a9e686d95e4df2f476aad8700906ed0/packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js#L99-L99), that uses the editing mode value from the store.

That means, if I leave the site editor or template preview mode after having activated "Write Mode", the post editor knows about it and won't render handy settings such as "Add before" and "Add after":

| Write mode enabled | Default |
|--------|--------|
| <img width="300" alt="Screenshot 2024-11-21 at 4 47 08 pm" src="https://github.com/user-attachments/assets/cfc1900e-7763-4f1f-a4ca-29c0ae5f2458"> | <img width="300" alt="Screenshot 2024-11-21 at 4 46 35 pm" src="https://github.com/user-attachments/assets/1b64bf35-2f6d-4bb4-aa80-6492768f464c"> | 

Especially swapping between site and post editors, I'd have expected to land in default editing mode for regular posting editing, and therefore have access to the block settings features.

talldan commented:
> That means, if I leave the site editor or template preview mode after having activated "Write Mode", the post editor knows about it and won't render handy settings such as "Add before" and "Add after":

I think this might be a bug in trunk at the moment. https://github.com/WordPress/gutenberg/issues/67084 tracks it for zoom out, but it also affects edit mode. I've also noticed it when editing patterns, you can be stuck in edit mode or zoom out mode. It's a pretty bad one!

talldan committed:
Handle when the SET_HAS_CONTROLLED_INNER_BLOCKS block has been removed from the state

aaronrobertshaw reviewed (commented):
Appreciate the detailed test instructions here @talldan 🙇 

For the most part, this PR tests as advertised. I also noticed the issue Ramon flagged but as you note that appears to be a separate issue.

There was one error I ran into though. If I create a pattern within the editor, via selecting multiple blocks and choosing "Create pattern", then click "Edit original" in the pattern block's toolbar, I get a console error regarding an attempt to access `clientId` on `undefined`.


https://github.com/user-attachments/assets/c36b2b64-f849-431d-8074-a74bc9f0e58f

<img width="942" alt="Screenshot 2024-11-21 at 4 36 13 pm" src="https://github.com/user-attachments/assets/c95afc99-315d-46fe-a59c-df620308aab7">


If I avoid triggering this error, the rest of the manual testing works as expected.

✅ If not in zoom out mode, I could only edit the overridden paragraph
✅ With zoom out mode enabled, the overridden paragraph wasn't editable
✅ After leaving zoom out mode and enabling write mode, the overridden paragraph is editable again
✅ Could not edit the nested pattern's overridden content

aaronrobertshaw commented:
> There was one error I ran into though

I restarted testing after the earlier rebase but looks like there was another commit since then. After pulling the latest and testing again I'm not able to replicate the earlier error. 

It's possible I'm doing something slightly different this time around so might need confirmation that nobody else can replicate the error.

talldan commented:
> I restarted testing after the earlier rebase but looks like there was another commit since then. After pulling the latest and testing again I'm not able to replicate the earlier error.

I did fix something along those lines in the last commit 👍 

I'll do some testing to make sure it was that. Thanks for the testing and detailed report.

aaronrobertshaw commented:
> I did fix something along those lines in the last commit 👍

If I understand the last commit, it makes sense it would fix it. The process I used to create the pattern would remove blocks and there's now a check to catch that scenario. 

I can replicate the issue with https://github.com/WordPress/gutenberg/pull/67026/commits/63d0d2f1ccf5999af0207272bc42e6c102ae927c checked out and it appears resolved when on https://github.com/WordPress/gutenberg/pull/67026/commits/773c76da031e510041fce1f3f96d00206ae1c3cf.

ramonjd reviewed (approved):
For what it's worth I took this for a spin in the post and site editors. Tested according to instructions in a couple of block themes. 

Pattern overrides are not editable in zoom out mode, but are in Write mode.

Nested patterns with overrides are not editable. 

Smoke tested editing templates, patterns and posts with write and zoom modes toggled on and off.

If other folks are good with it, then LGTM

talldan committed:
Inline editing mode calculation for individual blocks

talldan committed:
Calculate both the regular and nav mode derived block editing modes at the same time

talldan committed:
Update getEnabledClientIdsTree dependencies

talldan committed:
Fix tests

talldan committed:
Update more tests

talldan committed:
Enable write mode experiment for e2e pattern overrides block editing mode tests

talldan commented:
@youknowriad [As per this discussion](https://github.com/WordPress/gutenberg/pull/67026#discussion_r1849824177), I've updated this now so that it calculates both `derivedBlockEditingModes` and `derivedNavModeBlockEditingModes` at the same time (I'm not sure the names are the best, but seems like a minor thing).

Also addressed some of the other review feedback. Hopefully it's good to go now.

talldan commented:
I'm going to merge this seeing as nobody is telling me not to 😄 

The gutenberg RC was only just cut yesterday, so there should be a bit of time to resolve any issues or even revert again if needed.

ramonjd commented:
> I'm going to merge this seeing as nobody is telling me not to 😄

WAIT!

Just jokes.

🚢 

> The gutenberg RC was only just cut yesterday, so there should be a bit of time to resolve any issues or even revert again if needed before it lands in the next RC.

Sounds good. If there's anything, hopefully it'll pop up in the next two weeks with all the eyes on it.

priethor commented:
[Some of the changes to an e2e test](https://github.com/WordPress/gutenberg/pull/67232/files#diff-6bbe648f5452a157d4fd97c09ec3ed559f7b6665da1d3c9f6ed55af8c162987b) in #67232, which needs to be backported to the 19.8 branch, depend on this PR, so we have two options:
- Remove the e2e test from the 19.8 branch
- Backport this bugfix as well.

> If there's anything, hopefully it'll pop up in the next two weeks with all the eyes on it.

Since this is a bugfix, I'm tempted to backport it instead of removing the test. How confident are you about launching this tomorrow, @ramonjd @talldan ? 😄

talldan commented:
@priethor This PR is the one that introduced the E2E test in question, so removing the test from 19.8 seems like the best option to me.

This PR is pretty big, so I personally wouldn't want to see it cherry-picked into the release. There have been a couple of follow-up bug fixes it to it already, so it'd mean cherry-picking a few things.

priethor commented:
Thanks for confirming, Dan, I was leaning toward that as well 👍
